sys/defs/src/FreeMonoidX.hs

```diff
@@ -0,0 +1,298 @@
+--
+-- Can be written as
+--
+-- [a0, a1, ..., an-1]
+--
+-----------------------------------------------------------------------------
+
+module FreeMonoidX
+  ( -- * Free Monoid with multiplication Type
+    FreeMonoidX (..)
+
+  -- * Query
+  , nrofDistinctTerms
+  , distinctTerms  -- exposed for performance reasons checking properties for
+                   -- all distinct terms is faster than for all terms
+  , distinctTermsT
+  -- * Map
+  , mapTerms
+
+  -- * Filter
+  , partition
+  , partitionT
+
+  -- | Folds
+  , foldMultiplier
+  , foldFMX
+
+    -- * Manipulation of the free monoid
+  , append
+  , remove
+  , addNTimes
+  , flatten
+
+  -- * Multiplication operator
+  , (<.>)
+
+  -- * Monoid terms restriction
+  , IntMultipliable
+  , TermWrapper (..)
+
+  -- * Multiplier lists
+  , toMultiplierList
+  , toDistinctAscMultiplierList
+  , fromMultiplierList
+  , fromDistinctAscMultiplierList
+  -- ** Multiplier lists of `TermWrapper`'s
+  , fromListT
+  , toMultiplierListT
+  , toDistinctAscMultiplierListT
+  , fromMultiplierListT
+  , fromDistinctAscPowerListT
+  )
+where
+
+import           Control.Arrow   (first, (***))
+import           Control.DeepSeq
+import           Data.Foldable
+import           Data.List       hiding (partition)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Monoid     hiding (Product (..))
+import           GHC.Exts
+import           GHC.Generics    (Generic)
+
+-- | Symbolic representation of a polynomial, where each term is a member of
+-- type `a`.
+--
+-- The integer value of the map represents the number of occurrences of a term
+-- in the polynomial. Given this representation it is crucial that the
+-- operation is commutative, since the information of about the order of the
+-- term is lost in this representation.
+--
+newtype FreeMonoidX a = FMX { asMap :: Map a Integer }
+    deriving (Eq, Ord, Read, Generic, NFData)
+
+instance Show a => Show (FreeMonoidX a) where
+    show (FMX p) = show (Map.assocs p)
+
+-- | A term of the monoid which wraps a value. This could be for instance a
+-- sum-term or a product term.
+class TermWrapper f where
+    wrap :: a -> f a
+    unwrap :: f a -> a
+
+-- | Types that can be multiplied by an integral. This restriction is required
+-- to be able to assign a correct semantic to the symbolic representation of
+-- the free-monoid. If the elements of the free-monoid are not
+-- `IntMultipliable`, then `foldFMX` is ill-defined.
+--
+-- See the test code `FreeMonoidXspec` for examples of instances of this class.
+class IntMultipliable a where
+    -- | `n <.> x` multiplies `x` `n` times.
+    (<.>) :: Integral n
+         => n -- ^ Multiplication factor.
+         -> a -- ^ Element to multiply.
+         -> a
+
+instance Ord a => IntMultipliable (FreeMonoidX a) where
+    0 <.> _ = mempty
+    n <.> (FMX p) = FMX $ (toInteger n *) <$> p
+
+instance Ord a => Monoid (FreeMonoidX a) where
+    mempty = FMX []
```

Comment:
Why FMX [], while on next line FMX Map...?
list is not a map?

Reply:
I'm using the `OverloadedLists` extension.